### PR TITLE
[SPARK-20431][SQL] Specify a schema by using a DDL-formatted string

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr supported a DDL-formatted string in `DataFrameReader.schema`.
This fix could make users easily define a schema without importing  `o.a.spark.sql.types._`.

## How was this patch tested?
Added tests in `DataFrameReaderWriterSuite`.

### CODE REVIEW

**Comment by maropu** [comment]
cc: @gatorsmile

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -96,14 +96,17 @@ def schema(self, schema):
...
         spark = SparkSession.builder.getOrCreate()
-        jschema = spark._jsparkSession.parseDataType(schema.json())
-        self._jreader = self._jreader.schema(jschema)
+        if isinstance(schema, StructType):
+            jschema = spark._jsparkSession.parseDataType(schema.json())
+            self._jreader = self._jreader.schema(jschema)
+        elif isinstance(schema, basestring):
+            self._jreader = self._jreader.schema(schema)
+        else:
+            raise TypeError("schema should be StructType")
```
Comment (gatorsmile) [inline python/pyspark/sql/readwriter.py]: Update this message?

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 2]
```diff
@@ -68,6 +68,18 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
...
 
   /**
+   * Specifies the schema by using the input DDL-formatted string. Some data sources (e.g. JSON) can
+   * infer the input schema automatically from data. By specifying the schema here, the underlying
+   * data source can skip the schema inference step, and thus speed up data loading.
+   *
+   * @since 2.3.0
+   */
+  def schema(schemaString: String): DataFrameReader = {
+    this.userSpecifiedSchema = Option(StructType.fromDDL(schemaString))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: This change will make PySpark API inconsistent with the Scala API

File: python/pyspark/sql/readwriter.py (hunk 1 above)
Comment (maropu) [inline python/pyspark/sql/readwriter.py]: yea, I'll do

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala (hunk 2 above)
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Sorry, but I probably missed your point. What's the API consistency you pointed out here?
I just made the python APIs the same with the Scala ones like:
```
--- python
>>> from pyspark.sql.types import *
>>> fields = [StructField('a', IntegerType(), True), StructField('b', StringType(), True), StructField('c', DoubleType(), True)]
>>> schema = StructType(fields)
>>> spark.read.schema(schema).csv("/Users/maropu/Desktop/test.csv").show()
+---+----+---+
|  a|   b|  c|
+---+----+---+
|  1| aaa|0.3|
+---+----+---+

>>> spark.read.schema("a INT, b STRING, c DOUBLE").csv("/Users/maropu/Desktop/test.csv").show()
+---+----+---+
|  a|   b|  c|
+---+----+---+
|  1| aaa|0.3|
+---+----+---+

--- scala
scala> import org.apache.spark.sql.types._
scala> fields = StructField("a", IntegerType) :: StructField("b", StringType) :: StructField("c", DoubleType) :: Nil
scala> val schema = StructType(fields)
scala> spark.read.schema(schema).csv("/Users/maropu/Desktop/test.csv").show
+---+----+---+
|  a|   b|  c|
+---+----+---+
|  1| aaa|0.3|
+---+----+---+

scala> spark.read.schema("a INT, b STRING, c DOUBLE").csv("/Users/maropu/Desktop/test.csv").show
+---+----+---+
|  a|   b|  c|
+---+----+---+
|  1| aaa|0.3|
+---+----+---+
```

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: Sorry, I misread the Python codes.

**Comment by maropu** [comment]
@gatorsmile ping

**Comment by maropu** [comment]
@gatorsmile ping

File: python/pyspark/sql/readwriter.py [hunk 3]
```diff
@@ -96,14 +96,17 @@ def schema(self, schema):
         By specifying the schema here, the underlying data source can skip the schema
         inference step, and thus speed up data loading.
 
-        :param schema: a :class:`pyspark.sql.types.StructType` object
+        :param schema: a :class:`pyspark.sql.types.StructType` object or a DDL-formatted string
```
Comment (gatorsmile) [inline python/pyspark/sql/readwriter.py]: Could you give an example here to users?

**Comment by gatorsmile** [comment]
In PySpark, we have multiple ways to specify the schema. However, it sounds like we silently pick one of the input schemas. Could you submit a separate PR to fix the potential schema conflicts in PySpark?

Thanks!

File: python/pyspark/sql/readwriter.py (hunk 3 above)
Comment (maropu) [inline python/pyspark/sql/readwriter.py]: ok

**Comment by maropu** [comment]
Sure, I'd love to do tough, I probably missed your point. What's the scenario of the conflict you described? [The current logic in `readwriter.py`](https://github.com/maropu/spark/blob/46994fb2f22135f19ab615265cbe9da8f24aaa15/python/pyspark/sql/readwriter.py#L104) just checks types, then decides a schema. Could you give me a concrete example? Thanks

**Comment by gatorsmile** [comment]
nvm. I will fix it later. Thanks!

**Comment by gatorsmile** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Pull request closed without merging**